[RFC] Add `#[export_ordinal(n)]` attribute

MeguminSama opened the pull request:
This RFC adds an attribute that allows specifying the ordinal of an exported function when creating cdylibs.

We already have a [`#[link_ordinal(n)]`](https://doc.rust-lang.org/reference/items/external-blocks.html#the-link_ordinal-attribute) attribute which allows importing functions from DLLs by its ordinal, but we don't have any way of doing the opposite - specifying the ordinal of a function when we're making our own DLLs.

Currently, you need to create a `lib.def` file and specify the ordinals, and then link it with `cargo:rustc-cdylib-link-arg=/DEF`.

The biggest downside of the current method is that once you specify a `.def` file, you will have to specify an ordinal for every function that you want to export from the DLL, or else it won't be present in the generated `.lib` file. This can become very overwhelming if you have a lot of exported functions.

This RFC would allow you to specify exported function ordinals like so:

```rs
// Export this function at Ordinal 1
#[no_mangle]
#[export_ordinal(1)]
pub extern "C" fn hello() {
	println!("Hello, World!");
}
```

[Rendered](https://github.com/MeguminSama/rfcs/blob/export-function-ordinals/text/0000-export-function-ordinals.md)

programmerjake commented:
you should probably specify that the attribute is ignored on non-Windows (or similar) targets.

joshtriplett commented:
@rfcbot merge

tmandry commented:
Can we reuse `#[link_ordinal(n)]` for this purpose, the same way `extern "C"` can be used both for "importing" and "exporting" functions? I don't think there's any ambiguity based on context.

Should this attribute be unsafe once [unsafe attributes](https://rust-lang.github.io/rfcs/3325-unsafe-attributes.html) are implemented? I'm unsure if we can guarantee there will be no collisions with anything else linked into the program (including native libraries).

MeguminSama commented:
> Can we reuse `#[link_ordinal(n)]` for this purpose, the same way `extern "C"` can be used both for "importing" and "exporting" functions?

I think this could be a good idea if there is no ambiguity. It could be confusing at a glance if a project has multiple `#[link_ordinal(n)]`, one for importing, and one for exporting, but it's probably a non-issue.

> Should this attribute be unsafe once [unsafe attributes](https://rust-lang.github.io/rfcs/3325-unsafe-attributes.html) are implemented?

I'm not sure about how export_ordinal would be implemented, but if there's the chance of collisions, then I agree it should be marked as unsafe.

tmandry commented on text/3641-export-function-ordinals.md:
````diff
@@ -0,0 +1,145 @@
+}
+```
+
+If `#[export_ordinal(n)]` is not provided above an exported function, the current behaviour should apply, where an unused ordinal is provided sequentially. For example, if ordinals `1` and `3` are specified with the `export_ordinal` attribute, and there are two other exported functions without specified ordinals, they should use ordinals `2` and `4`.
````
This is also listed as an unresolved question below. Maybe it's better to leave it unresolved, since I think it sounds difficult to implement in a strict sense.

MeguminSama replied:
Thanks, this was an oversight - I've removed this text and kept it in the unresolved questions

tmandry reviewed (commented):
Does the attribute itself modify symbol attributes such that the symbol is exported, or does it only modify the ordinal field? I'm having a hard time finding the docs for the rules determining when a function is exported are not. I think it has to do with whether they are public, which seems like it would have some drawbacks. I'm surprised by the example in the RFC:

```rs
#[link(name = "external_library")]
extern "C" {
    #[export_ordinal(2)]
    pub fn external_function();
}
```

In this case is the externally-declared function somehow being re-exported? Then it seems like we might need both `#[link_ordinal(n)]` and `#[export_ordinal(n)]` on the same function.

> I'm not sure about how export_ordinal would be implemented, but if there's the chance of collisions, then I agree it should be marked as unsafe.

Can you add this as an unresolved question to the RFC?

MeguminSama commented:
> In this case is the externally-declared function somehow being re-exported?

@tmandry I took into consideration that some libraries could possibly be statically linked, and the possibility that someone may want to re-export a function. If this isn't a valid concern I'm happy to remove it - I just tried to think of every possibility.

RalfJung commented:
Are reexports even possible?
Existing attributes like `link_name` use the same name for imports and exports so it would seem odd to use a different name here.

tmandry commented:
> Existing attributes like `link_name` use the same name for imports and exports so it would seem odd to use a different name here.

I don't think that's true; if I try to use it on an export I get a warning:

```rust
#[link_name = "actual_symbol_name"]
pub extern "C" fn name_in_rust() {}
```

```
warning: attribute should be applied to a foreign function or static
 --> src/lib.rs:1:1
  |
1 | #[link_name = "actual_symbol_name"]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2 | #[no_mangle]
3 | pub extern "C" fn name_in_rust() {}
  | ----------------------------------- not a foreign function or static
  |
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: `#[warn(unused_attributes)]` on by default
```

Regardless, agreed that we should be consistent: Either `link_name` and `link_ordinal` work for both imports and exports, or we would have `export_name` and `export_ordinal`. I still lean toward the former _unless_ it's possible in some executable format to have a re-exported function under a different name.

RalfJung commented:
Seems like I misremembered...  I thought there was a way to set the link name independent of the item name. Maybe there isn't.

ChrisDenton commented:
Were you thinking of [`export_name`](https://doc.rust-lang.org/stable/reference/abi.html#the-export_name-attribute)? E.g.

```rust
#[export_name = "exported_symbol_name"]
pub extern "C" fn name_in_rust() {}
```

MeguminSama commented:
If we already use `export_name`, I think it'd be best to stick with `export_ordinal`?

RalfJung commented:
Yeah, that makes sense.

tmandry commented:
@rfcbot reviewed

Agreed that we should stick with `export_ordinal` to be consistent with `export_name`.

MeguminSama commented:
I noticed that the edition guide for unsafe attributes got [merged](https://github.com/ehuss/edition-guide/blob/b80cba8af64a9c52d56f7081c764e5396e406f6c/src/rust-2024/unsafe-attributes.md).

It says that `export_name` must be marked as `unsafe`. I think that if this is the case, `export_ordinal` should definitely be unsafe. 

Should I update the RFC to remove it from the unanswered questions, and clarify that in 2024 edition it should be marked as unsafe?

RalfJung commented:
If it has the same risks as `export_name`, then yeah it should be unsafe.

MeguminSama commented:
It's entirely possible that a dependency could mark something as a duplicate ordinal, so I believe the same risks apply.

I'm not sure whether to word it as "in 2024 edition" or not though.

RalfJung commented:
I don't see a reason to make this unsafe only in some editions. We do that for old attributes solely due to backwards compatibility. New things should be properly unsafe from the start.

MeguminSama commented:
Thanks for the clarification. I'll work on the changes in a bit.

MeguminSama commented:
I have updated the RFC to mark `export_ordinal` as unsafe.

I have also removed the section on re-exporting bindings, as after testing, I've confirmed that this is not valid code.

joshtriplett commented:
This doesn't have any blocking concerns; it's just waiting on checkboxes. Ping @nikomatsakis @pnkfelix @scottmcm: does anyone want to check their box asynchronously?

MeguminSama commented:
Is there anything I could do to help move this along?

joshtriplett commented:
This still seems simple and straightforward, and fills a gap for the ability to define Windows libraries in Rust.

@rfcbot merge

ChrisDenton commented:
+1 from me. This allows for keeping the function pointer at a consistent offset in the [export table](https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#the-edata-section-image-only) across versions of a DLL. This is admittedly more niche then just leaving it up to the linker and then importing by name but it is a useful tool and it pairs well with the existing [`link_ordinal`](https://doc.rust-lang.org/reference/items/external-blocks.html#the-link_ordinal-attribute) attribute, as the OP says.
